nonmember_subscript_operator/main.tex

```diff
@@ -125,6 +126,68 @@ \subsection{Compatiblity layer}
 
 \end{lstlisting}
 
+\subsection{Interpolated access}
+Subscripting a random-access range with non-integral types can have interesting use cases,
+like interpolating between the elements at the nearest smaller and larger integral index of the subscript:
+\smallskip\begin{lstlisting}
+auto operator[](std::ranges::random_access_range auto r, double index) {
+    double integ;
+    const double frac = std::modf(index, &integ);
+    return std::lerp(r[integ], r[integ + 1], frac);
+}
+std::vector v{0.0, 1.0, 4.5, 7.8};
+double value = v[1.6]; // value == 3.1
+\end{lstlisting}
+However, adding such an overload in practice might be dangerous because the floating-point index might truncate to a range's integral subscript operator in case the floating point overload is not visible, or change the behavior of an existing floating point subscription relying on the truncation.
+This issue could be worked around using a custom type which cannot truncate into an integral, allowing to e.g. also specify the used interpolation:
+\smallskip\begin{lstlisting}
+std::vector v{0.0, 1.0, 4.5, 7.8};
+double value = v[1.6_linear]; // value == 3.1
+\end{lstlisting}
+
+With the adoption of P2128, we could even allow interpolation on higher dimensional objects, like the proposed `std::mdspan` or `std::mdarray`.
+\smallskip\begin{lstlisting}
+std::mdarray<double, N, M, K> a = ...;
+double value = a[7.5, 8.9, 76.4]; // trilinear interpolation
+\end{lstlisting}
+
+While we are aware that such a usage might be niche or a separate function for interpolated access could be arguably better, we would like to give library writes the necessary machinery to implement such a feature within their domains if they desire so.
+
+\subsection{Generic subsetting}
+The idea from section \ref{sec:simd_load} to use SIMD vectors as subscripts can be further generalized to allow any kind entity representing multiple indices as subscript.
+\smallskip\begin{lstlisting}
+auto operator[](std::ranges::random_access_range r, std::ranges::input_range indices);
+\end{listing}
+The concrete implementation and return type of such an operation is debatable and thus intentionally omitted.
+Still, it could allow for better formulation of some algorithms or more efficient access for certain ranges, if the index set used for subscription is known in advance.
```

Comment:
Sure: we can always replace any operator by a function :) The question is which syntax feels more "natural". Maybe I should think more in a mathematical direction, where subscripting happens with non-integers.

The ADL part is a fair point! If I template the non-member `operator[]` on the first argument, I cannot be found if it is in another namespace. That makes it useless for `std` containers, since I cannot put the operator in namespace `std`. This is no obstacle for @mattkretz, because `simd` will live in `std`. Which makes me wonder where else people add non-member operators outside the namespaces of their parameter types.